domain/application/state/state.go

```diff
@@ -0,0 +1,279 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3"
+
+	coredb "github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/domain/application"
+	applicationerrors "github.com/juju/juju/domain/application/errors"
+	"github.com/juju/juju/domain/life"
+)
+
+// Logger is the interface used by the state to log messages.
+type Logger interface {
+	Debugf(string, ...interface{})
+}
+
+// State describes retrieval and persistence methods for storage.
+type State struct {
+	*domain.StateBase
+	logger Logger
+}
+
+// NewState returns a new state reference.
+func NewState(factory coredb.TxnRunnerFactory, logger Logger) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+		logger:    logger,
+	}
+}
+
+// UpsertApplication creates or updates the specified application,
+// also adding any units if specified.
+// TODO - this just creates a minimal row for now.
+func (s *State) UpsertApplication(ctx context.Context, name string, units ...application.AddUnitParams) error {
```

Comment:
I don't know if the variadic args here are intended to reflect the API server multi-params, but we shouldn't do it where where we supply individual results in the returned payload object - we should have a transaction for each and return success/error in each case.